File: bindings/python/tests/test_search.py
"""Focused end-to-end tests for the aysearch JSON binding."""

from aysearch import Model, SearchError


def test_search_spec_is_the_stable_safe_schema():
    model = Model("router")
    route = model.choice("route", ["cpu", "gpu"])
    cost = model.int("cost", 0, 20)
    model.table([route, cost], [["cpu", 3], ["gpu", 7]])
    model.add("cost <= 10")
    model.minimize(cost)

    assert model.to_spec() == {
        "version": 1,
        "name": "router",
        "variables": [
            {
                "name": "route",
                "domain": {"min": 0, "max": 1},
                "labels": {"0": "cpu", "1": "gpu"},
            },
            {"name": "cost", "domain": {"min": 0, "max": 20}},
        ],
        "constraints": [
            {"table": {"variables": ["route", "cost"], "tuples": [[0, 3], [1, 7]]}},
            {"expression": "cost <= 10"},
        ],
        "objective": {"sense": "minimize", "expression": "cost"},
    }


def test_search_solves_4x4_sudoku():
    model = Model("sudoku")
    cell = model.int_grid("cell", 4, 4, 1, 4)
    for row in cell:
        model.all_different(row)
    for column in range(4):
        model.all_different(cell[row][column] for row in range(4))
    for box_row in (0, 2):
        for box_column in (0, 2):
            model.all_different(
                cell[row][column]
                for row in range(box_row, box_row + 2)
                for column in range(box_column, box_column + 2)
            )
    for row, column, value in (
        (0, 0, 1),
        (0, 3, 4),
        (1, 1, 4),
        (2, 2, 4),
        (3, 0, 4),
        (3, 3, 1),
    ):
        model.add(f"{cell[row][column]} == {value}")

    result = model.solve()
    solution = result.require_solution()
    expected = {1, 2, 3, 4}
    assert all({solution[value] for value in row} == expected for row in cell)
    assert "declare-const |cell_0_0|" in model.to_smt2()


def test_search_choice_labels_and_complete_enumeration():
    model = Model("tiny router")
    route = model.choice("route", ["cpu", "gpu"])
    model.add("route >= 0")

    result = model.enumerate(10)
    assert result.status == "complete"
    assert result.complete is True
    assert {solution[route] for solution in result.solutions} == {"cpu", "gpu"}


def test_search_element_and_objective_enumeration_guard():
    model = Model("element")
    index = model.int("index", 0, 1)
    first = model.int("first", 4, 4)
    second = model.int("second", 9, 9)
    selected = model.int("selected", 4, 9)
    model.element(index, [first, second], selected)
    model.add("index == 1")
    assert model.solve().require_solution()[selected] == 9

    model.minimize(selected)
    try:
        model.enumerate()
    except SearchError as error:
        assert "objective" in str(error)
    else:
        raise AssertionError("enumerate() silently optimized an objective model")


def test_solution_rejects_a_variable_from_another_model():
    first = Model("first")
    own = first.int("same_name", 3, 3)
    solution = first.solve().require_solution()

    second = Model("second")
    foreign = second.int("same_name", 3, 3)
    try:
        solution[foreign]
    except SearchError as error:
        assert "another model" in str(error)
    else:
        raise AssertionError("solution accepted a variable from another model")

    assert solution[own] == 3
    assert solution["same_name"] == 3

File: bindings/js/test-search.mjs
import assert from "node:assert/strict";
import { Model, NativeSearchError, SearchError } from "./search.mjs";

const model = new Model("4x4 Sudoku");
const cell = model.intGrid("cell", 4, 4, 1, 4);
for (const row of cell) model.allDifferent(row);
for (let column = 0; column < 4; column++) {
  model.allDifferent(cell.map((row) => row[column]));
}
for (const boxRow of [0, 2]) {
  for (const boxColumn of [0, 2]) {
    model.allDifferent(
      [0, 1].flatMap((dr) =>
        [0, 1].map((dc) => cell[boxRow + dr][boxColumn + dc]),
      ),
    );
  }
}
for (const [row, column, value] of [
  [0, 0, 1],
  [0, 3, 4],
  [1, 1, 4],
  [2, 2, 4],
  [3, 0, 4],
  [3, 3, 1],
])
  model.add(`${cell[row][column]} == ${value}`);

const result = model.solve();
const solution = result.requireSolution();
for (const row of cell) {
  assert.deepEqual(
    new Set(row.map((value) => solution.get(value))),
    new Set([1, 2, 3, 4]),
  );
}
assert.match(model.toSMT2(), /declare-const \|cell_0_0\|/);

const choices = new Model("choices");
const route = choices.choice("route", ["cpu", "gpu"]);
choices.add("route >= 0");
const routes = choices.enumerate(10);
assert.equal(routes.status, "complete");
assert.deepEqual(
  new Set(routes.solutions.map((item) => item.get(route))),
  new Set(["cpu", "gpu"]),
);

const elementModel = new Model("element");
const index = elementModel.int("index", 0, 1);
const first = elementModel.int("first", 4, 4);
const second = elementModel.int("second", 9, 9);
const selected = elementModel.int("selected", 4, 9);
elementModel.element(index, [first, second], selected).add("index == 1");
assert.equal(elementModel.solve().requireSolution().get(selected), 9);
elementModel.minimize(selected);
assert.throws(() => elementModel.enumerate(), /objective/);

const emptyTable = new Model("empty table");
const emptyTableVariable = emptyTable.int("x", 0, 1);
assert.throws(() => emptyTable.table([emptyTableVariable], []), SearchError);

const prototypeNames = new Model("prototype-safe names");
const proto = prototypeNames.int("__proto__", 7, 7);
const constructor = prototypeNames.int("constructor", 11, 11);
assert.equal(Object.isFrozen(proto), true);
assert.throws(() => {
  proto.name = "renamed";
}, TypeError);
const prototypeSolution = prototypeNames.solve().requireSolution();
assert.equal(prototypeSolution.get(proto), 7);
assert.equal(prototypeSolution.get("__proto__"), 7);
assert.equal(prototypeSolution.rawValues.__proto__, 7);
assert.equal(prototypeSolution.get(constructor), 11);
assert.equal(Object.getPrototypeOf(prototypeSolution.values), null);
assert.equal(Object.getPrototypeOf(prototypeSolution.rawValues), null);
assert.equal(Object.hasOwn(prototypeSolution.values, "__proto__"), true);

const otherPrototypeNames = new Model("other model with the same names");
const foreignProto = otherPrototypeNames.int("__proto__", 7, 7);
assert.throws(() => prototypeSolution.get(foreignProto), /another model/);
assert.equal(prototypeSolution.get("__proto__"), 7); // String lookup stays valid.

const laterVariable = prototypeNames.int("declared_later", 13, 13);
assert.throws(() => prototypeSolution.get(laterVariable), /solve epoch/);
assert.equal(prototypeSolution.get(proto), 7);

const unsafeObjective = new Model("unsafe JavaScript objective");
const maxSafe = unsafeObjective.int(
  "max_safe",
  Number.MAX_SAFE_INTEGER,
  Number.MAX_SAFE_INTEGER,
);
unsafeObjective.maximize(`${maxSafe} + 1`);
assert.throws(() => unsafeObjective.solve(), NativeSearchError);

console.log("AY Search Node binding: PASS");
